SingularityUI/SingularityTailer/src/components/Log.js

```diff
@@ -0,0 +1,296 @@
+      } else {
+        // we are at the top/bottom, load some stuff.
+        if (nearTop && !this.isLineLoaded(0)) {
+          // don't dispatch in the raf, do it later
+          setTimeout(() => this.loadLine(0, true), 0);
+        }
+
+        if (nearBottom && lines.size) {
+          // if we haven't reached the end of the file yet
+          if (lines.last().isMissingMarker) {
+            // don't dispatch in the raf, do it later
+            setTimeout(() => this.loadLine(lines.size - 1, false), 0);
+          } else if (atBottom) {
+            // we're tailing here.
+            if (!this.state.tailing) {
+              this.setState({
+                tailing: true
+              });
+            }
+          }
+        }
+      }
+
+      if (!atBottom && this.state.tailing) {
+        this.setState({
+          tailing: false
+        });
+      }
+      // update the scroll position.
+      this.scrollTop = domNode.scrollTop;
+      this.scrollHeight = domNode.scrollHeight;
+    }
+    // do another raf.
+    this.rafRequestId = window.requestAnimationFrame(this.pollScroll);
+  }
+
+  loadLine(index, loadUp) {
+    return this.props.loadLine(
+      index,
+      loadUp,
+      this.props.lines,
+      this.props.chunks
+    );
+  }
+
+  tailLog() {
+    return this.props.tailLog(this.props.lines);
+  }
+
+  render() {
+    const { props } = this;
+
+    const hrefFunc = props.hrefFunc
+      ? (offset) => props.hrefFunc(props.tailerId, offset)
+      : undefined;
+
+    const logPaneClasses = classNames({
+      'log-pane': true,
+      tailing: this.state.tailing
+    });
+
+    return (
+      <section className={logPaneClasses}>
+        <div className="log-line-wrapper">
+          <LogLines
+            isLoaded={props.isLoaded}
+            lines={props.lines}
+            fakeLineCount={this.fakeLineCount}
+            isLineLoaded={this.isLineLoaded}
+            hrefFunc={hrefFunc}
+            lineLinkRenderer={props.lineLinkRenderer}
+            highlightedOffset={props.goToOffset}
+          />
+        </div>
+      </section>
+    );
+  }
+}
+
+Log.propTypes = {
+  tailerId: PropTypes.string.isRequired,
+  goToOffset: PropTypes.number,
+  hrefFunc: PropTypes.func,
+  minTailDelayMs: PropTypes.number.isRequired,
+  maxTailDelayMs: PropTypes.number.isRequired,
+  calculateTailDelayMs: PropTypes.func.isRequired,
+  // from connectToTailer HOC
+  getTailerState: PropTypes.func.isRequired,
+  // from tailer implementation
+  // actions
+  initializeFile: PropTypes.func.isRequired,
+  loadLine: PropTypes.func.isRequired,
+  tailLog: PropTypes.func.isRequired,
+  // from connect
+  isLoaded: PropTypes.bool.isRequired,
+  fileSize: PropTypes.number,
+  lines: PropTypes.instanceOf(Immutable.List),
+  chunks: PropTypes.instanceOf(Immutable.List),
+  requests: PropTypes.instanceOf(Immutable.Map),
+  config: PropTypes.object.isRequired,
+  lineLinkRenderer: PropTypes.func
+};
+
+Log.defaultProps = {
+  minTailDelayMs: 100,
```

Comment:
wondering if `100`ms is too low here. To a user, there probably wouldn't be much a perceptible difference between `100` and `500`. Concern here is the load on the api when hitting it every 100ms